mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp

```diff
@@ -686,21 +699,52 @@ LogicalResult ThreadwiseReadIntoRewritePattern::matchAndRewrite(
   Value bid = b.createOrFold<rock::WorkgroupIdOp>(loc, b.getIndexType());
   Value tid = b.createOrFold<rock::WorkitemIdOp>(loc, b.getIndexType());
 
-  SmallVector<Value, 3> readStartCoords = {bid, tid, zero};
+  SmallVector<Value, 3> readStartCoords{tid, zero};
+  SmallVector<int64_t, 3> bounds{1, numValues};
+  SmallVector<int64_t, 3> strides{1, vectorLen};
+  if (srcAddrSpace == gpu::AddressSpace::Global) {
+    readStartCoords.insert(readStartCoords.begin(), bid);
+    bounds.insert(bounds.begin(), 1);
+    strides.insert(strides.begin(), 1);
+  }
 
   auto loadLoop = b.create<TransformingForOp>(
       loc, ArrayRef<ValueRange>{readStartCoords, readStartCoords},
-      ArrayRef<Attribute>{transforms, b.getArrayAttr({})},
-      ArrayRef<int64_t>{1, 1, numValues}, ArrayRef<int64_t>{1, 1, vectorLen},
+      ArrayRef<Attribute>{transforms, b.getArrayAttr({})}, bounds, strides,
       forceUnroll, useIndexDiffs);
   {
     OpBuilder::InsertionGuard guard(b);
     b.setInsertionPointToStart(loadLoop.getBody());
-    Value loaded = b.create<GlobalLoadOp>(
-        loc, loadType, buffer, loadLoop.getValidity(/*domain=*/0),
-        loadLoop.getLowerCoords(/*domain=*/0));
-    b.create<InBoundsStoreOp>(loc, loaded, dest,
-                              loadLoop.getLowerCoords(/*domain=*/1)[2]);
+    if (srcAddrSpace == gpu::AddressSpace::Global) {
+      Value loaded = b.create<GlobalLoadOp>(
+          loc, loadType, buffer, loadLoop.getValidity(/*domain=*/0),
+          loadLoop.getLowerCoords(/*domain=*/0));
+      b.create<InBoundsStoreOp>(
+          loc, loaded, dest,
+          loadLoop.getLowerCoords(/*domain=*/1)[iterDim[srcAddrSpace]]);
+    } else {
+      TypedValue<IntegerType> valid = loadLoop.getValidity(/*domain=*/0);
+      Value zeroInt = createZeroConstantOp(b, loc, valid.getType());
+      Value isValid = b.createOrFold<arith::CmpIOp>(
+          loc, arith::CmpIPredicate::ne, valid, zeroInt);
+      scf::IfOp ifb =
+          b.create<scf::IfOp>(loc, isValid, /*withElseRegion=*/true);
+      {
+        OpBuilder thenb = ifb.getThenBodyBuilder(b.getListener());
+        Value loaded = thenb.create<InBoundsLoadOp>(
+            loc, loadType, buffer, loadLoop.getLowerCoords(/*domain=*/0));
+        thenb.create<InBoundsStoreOp>(
```

Comment:
Why not instead have the load be in the if, creating an if/else that yield's the value, and then unconditionally create the InBoundsStore?

Reply:
you are right. fixed now!